### PR TITLE
Migrate layers_names, line_alg and lars test to catch2

### PR DESCRIPTION
**Pull request by jeffin143**
Refer #2523 

Some more tests , Sorry for spamming :-p ---> `All tests passed (1354102 assertions in 910 test cases)` 

life-of-programmer

### CODE REVIEW

File: src/mlpack/tests/CMakeLists.txt [hunk 1]
```diff
@@ -232,7 +232,6 @@ set(parallel_tests
   "GMMTest;"
   "CFTest;"
   "HMMTest;"
-  "LARSTest;"
```
Comment (rcurtin) [inline src/mlpack/tests/CMakeLists.txt:235]: I think we will want to have this list of parallel tests back later; maybe that's worth adding to the issue description just so we don't forget?

File: src/mlpack/tests/lars_test.cpp [hunk 2]
```diff
@@ -40,17 +36,17 @@ void LARSVerifyCorrectness(arma::vec beta, arma::vec errCorr, double lambda)
     if (beta(j) == 0)
     {
       // Make sure that |errCorr(j)| <= lambda.
-      BOOST_REQUIRE_SMALL(std::max(fabs(errCorr(j)) - lambda, 0.0), tol);
+      REQUIRE(std::max(fabs(errCorr(j)) - lambda, 0.0) == Approx(0.0).margin(tol));
```
Comment (rcurtin) [inline src/mlpack/tests/lars_test.cpp:39]: This line is over 80 characters; can you wrap it? :)

File: src/mlpack/tests/lars_test.cpp [hunk 3]
```diff
@@ -298,16 +294,16 @@ BOOST_AUTO_TEST_CASE(TrainingAndAccessingBetaTest)
   LARS lars2;
   lars2.Train(X, y);
 
-  BOOST_REQUIRE_EQUAL(beta.n_elem, lars2.Beta().n_elem);
+  REQUIRE(beta.n_elem == lars2.Beta().n_elem);
   for (size_t i = 0; i < beta.n_elem; ++i)
-    BOOST_REQUIRE_CLOSE(beta[i], lars2.Beta()[i], 1e-5);
+    REQUIRE(beta[i] == Approx( lars2.Beta()[i]).epsilon(1e-7));
```
Comment (rcurtin) [inline src/mlpack/tests/lars_test.cpp:299]:
```suggestion
    REQUIRE(beta[i] == Approx(lars2.Beta()[i]).epsilon(1e-7));
```
:)

File: src/mlpack/tests/lars_test.cpp [hunk 4]
```diff
@@ -217,20 +213,20 @@ BOOST_AUTO_TEST_CASE(PredictRowMajorTest)
...
-  BOOST_REQUIRE_EQUAL(colMajorPred.n_elem, rowMajorPred.n_elem);
+  REQUIRE(colMajorPred.n_elem == rowMajorPred.n_elem);
   for (size_t i = 0; i < colMajorPred.n_elem; ++i)
   {
     if (std::abs(colMajorPred[i]) < 1e-5)
-      BOOST_REQUIRE_SMALL(rowMajorPred[i], 1e-5);
+      REQUIRE(rowMajorPred[i] == Approx(0.0).margin(1e-5));
     else
-      BOOST_REQUIRE_CLOSE(colMajorPred[i], rowMajorPred[i], 1e-5);
+      REQUIRE(colMajorPred[i] == Approx( rowMajorPred[i]).epsilon(1e-7));
```
Comment (rcurtin) [inline src/mlpack/tests/lars_test.cpp:222]:
```suggestion
      REQUIRE(colMajorPred[i] == Approx(rowMajorPred[i]).epsilon(1e-7));
```
:)

File: src/mlpack/tests/lars_test.cpp [hunk 5]
```diff
@@ -185,20 +181,20 @@ BOOST_AUTO_TEST_CASE(PredictTest)
...
-        BOOST_REQUIRE_EQUAL(predictions.n_elem, 1000);
+        REQUIRE(predictions.n_elem == 1000);
         for (size_t i = 0; i < betaOptPred.n_elem; ++i)
         {
           if (std::abs(betaOptPred[i]) < 1e-5)
-            BOOST_REQUIRE_SMALL(adjPred[i], 1e-5);
+            REQUIRE(adjPred[i] == Approx(0.0).margin(1e-5));
           else
-            BOOST_REQUIRE_CLOSE(adjPred[i], betaOptPred[i], 1e-5);
+            REQUIRE(adjPred[i] == Approx( betaOptPred[i]).epsilon(1e-7));
```
Comment (rcurtin) [inline src/mlpack/tests/lars_test.cpp:190]:
```suggestion
            REQUIRE(adjPred[i] == Approx(betaOptPred[i]).epsilon(1e-7));
```
:)

File: src/mlpack/tests/lars_test.cpp [hunk 6]
```diff
@@ -320,16 +316,16 @@ BOOST_AUTO_TEST_CASE(TrainingConstructorWithDefaultsTest)
 
   LARS lars2(X, y);
 
-  BOOST_REQUIRE_EQUAL(beta.n_elem, lars2.Beta().n_elem);
+  REQUIRE(beta.n_elem == lars2.Beta().n_elem);
   for (size_t i = 0; i < beta.n_elem; ++i)
-    BOOST_REQUIRE_CLOSE(beta[i], lars2.Beta()[i], 1e-5);
+    REQUIRE(beta[i] == Approx( lars2.Beta()[i]).epsilon(1e-7));
```
Comment (rcurtin) [inline src/mlpack/tests/lars_test.cpp:321]:
```suggestion
    REQUIRE(beta[i] == Approx(lars2.Beta()[i]).epsilon(1e-7));
```

File: src/mlpack/tests/lars_test.cpp [hunk 7]
```diff
@@ -347,15 +343,15 @@ BOOST_AUTO_TEST_CASE(TrainingConstructorWithNonDefaultsTest)
 
   LARS lars2(X, y, transposeData, useCholesky, lambda1, lambda2);
 
-  BOOST_REQUIRE_EQUAL(beta.n_elem, lars2.Beta().n_elem);
+  REQUIRE(beta.n_elem == lars2.Beta().n_elem);
   for (size_t i = 0; i < beta.n_elem; ++i)
-    BOOST_REQUIRE_CLOSE(beta[i], lars2.Beta()[i], 1e-5);
+    REQUIRE(beta[i] == Approx( lars2.Beta()[i]).epsilon(1e-7));
```
Comment (rcurtin) [inline src/mlpack/tests/lars_test.cpp:348]:
```suggestion
    REQUIRE(beta[i] == Approx(lars2.Beta()[i]).epsilon(1e-7));
```

File: src/mlpack/tests/layer_names_test.cpp [hunk 8]
```diff
@@ -63,69 +60,69 @@ BOOST_AUTO_TEST_CASE(LayerNameVisitorTest)
   // Bilinear interpolation is not yet supported by the string converter.
   LayerTypes<> unsupportedLayer = new BilinearInterpolation<>();
 
-  BOOST_REQUIRE(boost::apply_visitor(LayerNameVisitor(),
+  REQUIRE(boost::apply_visitor(LayerNameVisitor(),
                       atrousConvolution) == "atrousconvolution");
```
Comment (rcurtin) [inline src/mlpack/tests/layer_names_test.cpp:64]: I know this is tedious, but maybe you can wrap these second lines so they're all doubly-indented from the first?
```suggestion
      atrousConvolution) == "atrousconvolution");
```

File: src/mlpack/tests/lin_alg_test.cpp [hunk 9]
```diff
@@ -52,10 +50,13 @@ BOOST_AUTO_TEST_CASE(TestCenterA)
   //  [-8 -4  0  4  8]]
   for (int row = 0; row < 5; row++)
     for (int col = 0; col < 5; col++)
-      BOOST_REQUIRE_CLOSE(tmp_out(row, col), (double) (col - 2) * row, 1e-5);
+    {
+      REQUIRE(tmp_out(row, col) ==
+          Approx((double) (col - 2) * row).epsilon(1e-7));
+    }
```
Comment (rcurtin) [inline src/mlpack/tests/lin_alg_test.cpp:56]: Just another simple style comment; if the inner loop now has braces, can you also add braces to the outer loop?  (This is mentioned somewhere in the style guide.)

File: src/mlpack/tests/lin_alg_test.cpp [hunk 10]
```diff
@@ -75,10 +76,13 @@ BOOST_AUTO_TEST_CASE(TestCenterB)
   //  [-10  -6   -2   2   6   10 ]]
   for (int row = 0; row < 5; row++)
     for (int col = 0; col < 6; col++)
-      BOOST_REQUIRE_CLOSE(tmp_out(row, col), (double) (col - 2.5) * row, 1e-5);
+    {
+      REQUIRE(tmp_out(row, col) ==
+          Approx((double) (col - 2.5) * row).epsilon(1e-7));
+    }
```
Comment (rcurtin) [inline src/mlpack/tests/lin_alg_test.cpp:82]: Same comment here. :)

**Review by rcurtin — COMMENTED** [review]
Awesome, looks good @jeffin143!  I just have a bunch of tiny style comments. :)

File: src/mlpack/tests/CMakeLists.txt (hunk 1 above)
Comment (jeffin143) [reply src/mlpack/tests/CMakeLists.txt:235]: @zoq , Can you help us with this ?

Comment (zoq) [reply src/mlpack/tests/CMakeLists.txt:235]: I agree with @rcurtin, let's update the issue description mention that we should compare the test runtime with the boost runtime once the transition is complete, if the difference is huge let's find a solution to run parallel tests.

**Review by zoq — APPROVED** [review]
giphy

**Pull request merged**